jmxtrans-core/src/main/java/com/googlecode/jmxtrans/model/JmxResultProcessor.java

```diff
@@ -58,132 +58,137 @@ public JmxResultProcessor(Query query, ObjectInstance objectInstance, List<Attri
 	}
 
 	public ImmutableList<Result> getResults() {
-		Builder<Result> accumulator = ImmutableList.builder();
+		ResultsBuilder builder = new ResultsBuilder();
 		for (Attribute attribute : attributes) {
-			getResult(accumulator, attribute);
+			builder.add(attribute.getName(), attribute.getValue());
 		}
-		return accumulator.build();
+		return builder.build();
 	}
 
 	/**
-	 * Used when the object is effectively a java type
+	 * Result list builders.
+	 * Recursively walks in the value to add results.
 	 */
-	private void getResult(Builder<Result> accumulator, Attribute attribute) {
-		Object value = attribute.getValue();
-		if (value == null) {
-			return;
-		}
+	private class ResultsBuilder {
+		private final Builder<Result> accumulator = ImmutableList.builder();
+		private final long epoch = System.currentTimeMillis();
 
-		if (value instanceof CompositeData) {
-			getResult(accumulator, attribute.getName(), (CompositeData) value);
-		} else if (value instanceof CompositeData[]) {
-			for (CompositeData cd : (CompositeData[]) value) {
-				getResult(accumulator, attribute.getName(), cd);
+		/**
+		 * Add one or more results from a value of any type.
+		 * This is a recursive function.
+         */
+		private void add(String name, Object value) {
+			if (value == null) {
+				return;
 			}
-		} else if (value instanceof ObjectName[]) {
-			Map<String, Object> values = newHashMap();
-			for (ObjectName obj : (ObjectName[]) value) {
-				values.put(obj.getCanonicalName(), obj.getKeyPropertyListString());
+			if (value instanceof CompositeData) {
+				add(name, (CompositeData) value);
+			} else if (value instanceof CompositeData[]) {
+				for (CompositeData cd : (CompositeData[]) value) {
+					add(name, cd);
+				}
+			} else if (value instanceof ObjectName[]) {
+				add(name, (ObjectName[]) value);
+			} else if (value.getClass().isArray()) {
+				// OMFG: this is nutty. some of the items in the array can be
+				// primitive! great interview question!
+				Map<String, Object> values = newHashMap();
```

Comment:
* Array of ObjectName: add one Result with a special map
* Array of CompositeData: add Results for each CompositeData
* Array of anything else: add one Result with map similar to the array